homeassistant/components/ring/switch.py

```diff
@@ -0,0 +1,111 @@
+
+SIREN_ICON = 'mdi:alarm-bell'
+
+"""
+It takes a few seconds for the API to correctly return an update indicating
+that the changes have been made. Once we request a change (i.e. a light
+being turned on) we simply wait for this time delta before we allow
+updates to take place.
+"""
+SKIP_UPDATES_DELAY = timedelta(seconds=5)
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Create the switches for the Ring devices."""
+    cameras = hass.data[DATA_RING_STICKUP_CAMS]
+    switches = []
+
+    for device in cameras:
+        if device.has_capability('siren'):
+            switches.append(SirenSwitch(device))
+
+    add_entities(switches, True)
+
+
+class BaseRingSwitch(SwitchDevice):
+    """Represents a switch for controlling an aspect of a ring device."""
+
+    def __init__(self, device, device_type):
+        """Initialize the switch."""
+        self._device = device
+        self._device_type = device_type
+        self._unique_id = '{}-{}'.format(self._device.id, self._device_type)
+
+    async def async_added_to_hass(self):
+        """Register callbacks."""
+        async_dispatcher_connect(
+            self.hass, SIGNAL_UPDATE_RING, self._update_callback)
+
+    @callback
+    def _update_callback(self):
+        """Call update method."""
+        self.async_schedule_update_ha_state(True)
+        _LOGGER.debug("Updating Ring sensor %s (callback)", self.name)
+
+    @property
+    def name(self):
+        """Name of the device."""
+        return '{} {}'.format(self._device.name, self._device_type)
+
+    @property
+    def unique_id(self):
+        """Return a unique ID."""
+        return self._unique_id
+
+    @property
+    def should_poll(self):
+        """Updates controlled via the hub."""
+        return False
+
+
+class SirenSwitch(BaseRingSwitch):
+    """Creates a switch to turn the ring cameras siren on and off."""
+
+    def __init__(self, device):
+        """Initialize the switch for a device with a siren."""
+        super().__init__(device, 'siren')
+        self._no_updates_until = datetime.now()
+        self._siren_on = False
+
+    def __set_switch(self, new_state):
+        """Update switch state, and causes HASS to correctly update."""
+        self._device.siren = new_state
+        self._siren_on = new_state > 0
+        self._no_updates_until = datetime.now() + SKIP_UPDATES_DELAY
+        self.async_schedule_update_ha_state(True)
+
+    @property
+    def is_on(self):
+        """If the switch is currently on or off."""
+        return self._siren_on
+
+    def turn_on(self, **kwargs):
+        """Turn the siren on for 30 seconds."""
+        self.__set_switch(1)
+
+    def turn_off(self, **kwargs):
+        """Turn the siren off."""
+        self.__set_switch(0)
+
+    @property
+    def icon(self):
+        """Return the icon."""
+        return SIREN_ICON
+
+    def update(self):
+        """Update state of the siren."""
+        if self._no_updates_until > datetime.now():
```

Comment:
Why do we need to check for this? We have turned off automatic polling. `update` will only be called during entity addition.

Reply:
Because when the button is clicked it checks the state of the device which still returns it's off. It causes the switch to bounce on then off and wait for the normal update to happen. This causes it to "look" more correct (i.e. when you turn it on it goes on and stays on)

Reply:
This method `update` isn't called when `turn_on` or `turn_off` is called. I'm not following.

Reply:
Looks like we can remove this check if we don't force `update` to be called when updating state at the end of `set_switch`.

Reply:
The issue then is a background update can be called (and is every 10 seconds). It takes around 3 seconds to update so it was still happening frequently enough to be annoying on the UI. This totally protects from that, but I admit does make the code more complex. It's your call.

Reply:
Ok, let's keep it.